### PR TITLE
gh-103462: Ensure SelectorSocketTransport.writelines registers a writer when data is still pending

### LINKED ISSUE #103462: Calling SelectorSocketTransport.writelines with a very large payload doesn't get completely written to the socket
# Bug report

As a consequence of the optimization introduced in #91166, when `SelectorSocketTransport.writelines` is used with a very large buffer which can't be written in one shot either by using `socket.send` or `socket.sendmsg` the remaining data in the buffer is never written to the socket. 

The following example can be used to reproduce the error:

```python
import asyncio


async def send_message(chunk_size: int, chunks: int):
    await asyncio.sleep(1)
    reader, writer = await asyncio.open_connection("127.0.0.1", 9898)
    data = [bytes(chunk_size)] * chunks + [b"end"]
    writer.writelines(data)
    await writer.drain()
    writer.close()
    await writer.wait_closed()


async def handle_message(reader, writer):
    data = await reader.read()
    print(f"End marker: {data[-3:]!r}")
    writer.close()
    await writer.wait_closed()


async def run_server():
    server = await asyncio.start_server(handle_message, "127.0.0.1", 9898)
    async with server:
        await server.serve_forever()


async def main():
    await asyncio.gather(run_server(), send_message(6500, 100))


asyncio.run(main())
```

The example above prints `End marker: b'end'` with python 3.11 but not with python `3.12.0a6+`

# Your environment

- CPython versions tested on: `3.12.0a7`
- Operating system and architecture: macOS 13.2

### Linked PRs
* gh-103463

### PR DESCRIPTION
**Pull request by alisaifee**
# Description
When the payload used with `SelectorSocketTransport.writelines` is too large to write in one call to either `socket.send` or `socket.sendmsg` the remaining buffer needs to be handled by registering a write handler.

## Related Pull Request / Issue
- #91166 

* Issue: gh-103462

### CODE REVIEW

**Review by kumaraditya303 — COMMENTED** [review]
LGTM, Can you add a news entry? Thanks

**Comment by alisaifee** [comment]
> LGTM, Can you add a news entry? Thanks

done!

File: Misc/NEWS.d/next/Library/2023-04-12-06-00-02.gh-issue-103462.w6yBlM.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Fixed an issue with using :meth:`writelines` in :mod:`asyncio` to send very
+large payloads that exceed the amount of data that can be written in one
+call to :meth:`socket.socket.send` or :meth:`socket.socket.sendmsg`,
+resulting in the remaining buffer being left unwritten. This is fixed by
+registering a writer when the buffer still exists after the first write.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2023-04-12-06-00-02.gh-issue-103462.w6yBlM.rst]:
```suggestion
Fixed an issue with using :meth:`~asyncio.WriteTransport.writelines` in :mod:`asyncio` to send very
large payloads that exceed the amount of data that can be written in one
call to :meth:`socket.socket.send` or :meth:`socket.socket.sendmsg`,
resulting in the remaining buffer being left unwritten.
```

**Review by kumaraditya303 — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2023-04-12-06-00-02.gh-issue-103462.w6yBlM.rst (hunk 1 above)
Comment (alisaifee) [reply Misc/NEWS.d/next/Library/2023-04-12-06-00-02.gh-issue-103462.w6yBlM.rst]: Whooops! Thanks for fixing that!

**Pull request merged**